<!DOCTYPE html>
<html>
  <head>
    <meta charset='utf-8' />
    <meta http-equiv="X-UA-Compatible" content="chrome=1" />
    <meta name="description" content="SUPERSMART : Self-Updating Platform for the Estimation of Rates of Speciation, Migration And Relationships of Taxa" />
    <link rel="stylesheet" type="text/css" media="screen" href="stylesheets/stylesheet.css" />
    <link rel="icon" href="images/favicon.ico" />
    <script type="text/javascript" src="javascripts/jquery.js"></script>
    <script type="text/javascript" src="javascripts/main.js"></script>    
    <title>SUPERSMART</title>
  </head>
  <body>

	<!-- NAVIGATION -->
	<div id="navbar">
	<ul>
	  <li><a href="#intro" class="scrollable">Introduction</a></li>
	  <li><a href="#install" class="scrollable">Install</a></li>
	  <li><a href="#run" class="scrollable">Running</a></li>
	  <li><a href="#advanced" class="scrollable">Advanced topics</a></li>
	  <li><a href="#resources" class="scrollable">Resources</a></li>
	  <li><a href="#people" class="scrollable">People</a></li>	  
	  <li><a href="#acknowledgements" class="scrollable">Acknowledgements</a></li>	  	  
	</ul>
	</div>

    <!-- HEADER -->
    <div id="header_wrap" class="outer">
        <header class="inner">
          <a id="forkme_banner" href="https://github.com/naturalis/supersmart">View on GitHub</a>

          <img src="images/logo200x480.png" width="480" height="200" alt="supersmart logo" class="clean"/>
          <!--h1 id="project_title">SUPERSMART</h1>
          <h2 id="project_tagline">Self-Updating Platform for the Estimation of Rates of Speciation, Migration And Relationships of Taxa</h2-->

            <section id="downloads">
              <a class="zip_download_link" href="https://github.com/naturalis/supersmart/zipball/master">Download this project as a .zip file</a>
              <a class="tar_download_link" href="https://github.com/naturalis/supersmart/tarball/master">Download this project as a tar.gz file</a>
            </section>
        </header>
    </div>

    <!-- MAIN CONTENT -->
    <div id="main_content_wrap" class="outer">      
      
      <section id="main_content" class="inner">

      <!-- intro section -->
      <h2 id="intro" class="section"><a name="intro"></a>Introduction</h2>
      <p>
      	The Self-Updating Platform for Estimating Rates of Speciation and Migration,
      	Ages and Relationships of Taxa (SUPERSMART) is a self-contained, 
      	<a href="#install" class="scrollable">easy to install</a> analytical 
      	environment for large-scale phylogenetic data mining, taxonomic name resolution, 
      	tree inference and fossil-based tree calibration.
      </p>
      <img src="images/Fig_2_Approach-comparison.png" class="clean" width="640" height="530" alt="comparison" />
      <p>
      	Unlike supertree and supermatrix approaches, SUPERSMART decomposes the tree 
      	searching problem into subproblems so that complicated recent divergences can 
      	be reconstructed using the current state-of-the-art (the multispecies, multilocus
      	coalescent), while retaining the ability to build very large, composite estimates
      	of phylogeny with branch lengths proportional to time.
      </p>
      <p>
      	<strong>
      	Detailed information about SUPERSMART is available in 
      	<a href="http://dx.doi.org/10.7287/peerj.preprints.501v1">
      	10.7287/peerj.preprints.501v1</a>
      	</strong>
      </p>
      
      <!-- install section -->
      <h2 id="install" class="section"><a name="install"></a>Installation</h2>
      <p>
      	We have made every effort to make installation of SUPERSMART as simple as possible
      	on as many different computers as possible. The three, easy steps described below are 
      	known to work on the majority of modern computers with recent versions of Windows, 
      	Mac OS X and Linux-like operating systems. The end result is an installation that 
      	allows you to try out SUPERSMART, but it does not take full advantage of some of 
      	the speedups provided by the kind of high-performance computing (HPC) clusters 
      	that are available at certain academic computing centres<sup>*</sup>. 
      </p>
      <small>(<sup>*</sup>To install the HPC version of SUPERSMART go to the 
      <a href="#advanced" class="scrollable">advanced topics</a>.)</small>
      <h3>Step 1: Install VirtualBox</h3>
      <p>
      	<img src="images/virtualbox155x157-glow.png" class="clean inline" width="155" height="157" alt="virtualbox" />
      	VirtualBox extends the capabilities of your existing computer so that it can run 
      	multiple operating systems (inside multiple "virtual machines", or VMs) at the 
      	same time. Because SUPERSMART consists of many different tools that depend on 
      	one another we make it available as a VM that can run on any operating system
      	that VirtualBox can run on. VirtualBox is free and open source software. 
      	<a href="http://www.virtualbox.org/wiki/Downloads">Go here to download</a>, then
      	run the installer.
      </p>
      <h3>Step 2: Install Vagrant</h2>
      <p>
      	<img src="images/Vagrant155x139-glow.png" class="clean inline" width="155" height="139" alt="vagrant" />      
        Vagrant is a system for downloading, managing and launching virtual machines. It
        tightly integrates with VirtualBox to fetch and run the most up to date version
        of the SUPERSMART VM and to connect it to your existing computer (e.g. to exchange
        data). Vagrant is free and open source software.
        <a href="http://www.vagrantup.com">Go here to download</a>, then run the installer.
      </p>
      <h3>Step 3: Download and launch SUPERSMART</h3>
      <p>
		In the last step, 
		<a href="https://github.com/naturalis/supersmart/zipball/master">download SUPERSMART
		</a>itself, unzip the downloaded file (in some cases this happens automatically),
		navigate in the terminal to the unzipped folder and issue the following commands:
      </p>      
      <pre>
<span class="term">&gt;</span> cd supersmart-master/conf
<span class="term">&gt;</span> vagrant up
<span class="term">&gt;</span> vagrant ssh</pre>
	  <p>
	  	<strong>Please note:</strong> the <code>vagrant up</code> step will download a large amount
	  	of data (~8Gb). This is a one time operation. Ensure that you have a fast internet 
	  	connection and enough drive space for this. The <code>vagrant ssh</code> command
	  	will log you into the SUPERSMART environment.
	  </p>
      
      <!-- run section -->
      <h2 id="run" class="section"><a name="run"></a>Running</h2>
      	<p>
      		The SUPERSMART pipeline consists of a number of different steps that can be
      		chained together by issuing commands in a terminal window, or by using 
      		scripting tools such as shell scripts, Makefiles, and so on.
      	</p>      	
      	<h3>The <code>smrt</code> command</h3>
		  <p>
			When logged into the SUPERSMART environment, steps of the pipeline can be
			executed by issuing the <code>smrt</code> command in the 
			terminal with the appropriate global options, subcommands, and their arguments, 
			as follows:		
		  </p>
		  <pre>smrt [OPTIONS] [commands|help] [SUBCOMMAND] [ARGUMENTS]</pre>
		  <p>Examples of different ways in which <code>smrt</code> can be 
		  used:</p>
		  <dl>
			<dt>smrt</dt>
			<dd>Prints brief, overall help message and quits.</dd>	
			<dt>smrt --version</dt>
			<dd>Prints version number and quits.</dd>	  
			<dt>smrt commands</dt>
			<dd>Lists all available subcommands and quits.</dd>
			<dt>smrt help [SUBCOMMAND]</dt>
			<dd>Prints help message about the specified subcommand and quits.</dd>
			<dt>smrt [SUBCOMMAND] [ARGUMENTS]</dt>
			<dd>Runs the specified subcommand with the provided arguments.</dd>
			<dt>smrt [OPTIONS] [SUBCOMMAND] [ARGUMENTS]</dt>
			<dd>Runs the specified subcommand with the provided arguments and with 
			alternative global options in effect. The following global options are 
			available:
			  <dl>
				<dt>--workdir=&lt;location&gt;</dt>
				<dd>Specifies where data files are written to and read from. By default
				this is the current working directory.</dd>
				<dt>--verbose</dt>
				<dd>Sets the global verbosity, can be repeated multiple times. By default,
				verbosity is set such that warning and error messages are shown, 
				but informational and debugging messages are hidden.</dd>
			  </dl>		  	
			</dd>
		  </dl>  
		  <h3>Available subcommands</h3> 	
		  <p>The following subcommands, which represent steps of the pipeline, are 
		  available. For detailed usage, consult their respective usage messages by 
		  issuing <code>smrt help [SUBCOMMAND]</code>.</p>
		  <dl>
		  	<dt>taxize</dt>
		  	<dd>Performs taxonomic name resolution by mapping a provided list of input
		  	names onto the NCBI taxonomy. Optionally expands the higher input taxa to the 
		  	specified lower taxon, e.g. to expand a named Order to its constituent 
		  	species. Produces a table that lists all resolved (and expanded) taxa and 
		  	their higher classification.</dd>
		  	<dt>classify</dt>
		  	<dd>Given an input table of resolved taxa, produces a taxonomy tree, i.e.
		  	a polytomous cladogram with labeled interior nodes, in Newick format. This
		  	tree is used as a starting tree to infer a backbone phylogeny.</dd>
		  	<dt>align</dt>
		  	<dd>Given an input table of resolved taxa, performs multiple sequence 
		  	alignment of all potentially useful <a href="http://phylota.net">PhyLoTA</a>
		  	clusters for the taxa of interest. Produces a list of aligned candidate 
		  	clusters. The alignment method can be configured, effective methods 
		  	include those provided by <code>muscle</code> and <code>mafft</code>.</dd>
		  	<dt>orthologize</dt>
		  	<dd>Given a list of aligned candidate clusters, assigns orthology among the
		  	clusters by performing reciprocal <code>blast</code> searches on the seed 
		  	sequences around which the clusters were assembled. Produces a list of 
		  	re-aligned superclusters.</dd>
		  	<dt>bbmerge</dt>
		  	<dd>Given a list of superclusters and an input table of resolved taxa, creates
		  	a supermatrix that can be used to infer a backbone phylogeny. The supermatrix
		  	contains up to two, distal exemplar species from each genus and an optimized 
		  	selection of superclusters that minimizes sparseness of the supermatrix while
		  	attaining a user-specified minimum number of distinct markers for each taxon
		  	and omitting any alignments that exceed a user-specified maximum amount of
		  	sequence divergence.
		  	</dd>
		  	<dt>bbinfer</dt>
		  	<dd>Given a supermatrix and a starting tree, infers a backbone phylogeny. This
		  	backbone is typically a tree that reconstructs the relationships among a large
		  	number of genera (represented by up to two exemplar species) and so this step
		  	employs large-scale phylogenetic inference methods such as provided by 
		  	<code>ExaML</code> and similar tools.</dd>
		  	<dt>bbreroot</dt>
		  	<dd>Given a backbone phylogeny and a table of resolved taxa, reroots the 
		  	phylogeny to approximate the rooting implied by the classification hierarchy 
		  	of the provided, resolved taxa.</dd>
		  	<dt>bbcalibrate</dt>
		  	<dd>Given a rooted molecular backbone phylogeny and a table of fossils, 
		  	performs tree calibration using <code>treePL</code>. Produces an ultrametric 
		  	tree with branch lengths proportional to evolutionary time (i.e. a 
		  	"chronogram").</dd>
		  	<dt>bbdecompose</dt>
		  	<dd>Given a rooted backbone phylogeny, a list of superclusters and a table of 
		  	resolved taxa, decomposes the backbone into its constituent, most recent, 
		  	monophyletic clades, expands these clades into all taxa from the provided
		  	table and assembles sets of alignments for each clade that can be used for
		  	further tree inference within that clade.</dd>
		  	<dt>clademerge</dt>
		  	<dd>For each decomposed clade, merges the set of alignments assembled for this
		  	clade into an input file for tree inference.</dd>
		  	<dt>cladeinfer</dt>
		  	<dd>For each decomposed clade, infers the species tree. This is done using
		  	the multispecies, multilocus coalescent as implemented in <code>*BEAST</code>
		  	on the basis of the merged set of alignments for that clade.</dd>
		  	<dt>cladegraft</dt>
		  	<dd>Grafts the inferred clade trees on the backbone chronogram.</dd>
		  </dl>     
      
      
      
      <!-- advanced section -->
      <h2 id="advanced" class="section"><a name="advanced"></a>Advanced topics</h2>
      <h3>Installing on HPC systems</h3>
      <p>
      	To take advantage of the infrastructure provided by HPC clusters it is generally
      	best to compile key components of the SUPERSMART pipeline from source code. To 
		simplify this process we provide a "provisioning" script. This script is written
		for <a href="http://puppetlabs.com/puppet/puppet-open-source">puppet</a>, which 
		consequently must be available on your system. The script compiles and installs all 
		dependencies, data, and code that SUPERSMART needs. At present this is done in 
		system-wide locations, so the commands below must be 
		executed by a privileged user<sup>*</sup>:
	  </p>
		<pre>
# git clone https://github.com/naturalis/supersmart.git
# cd supersmart/conf
# puppet apply </pre>
      <p>
		The provisioning that commences when the <code>puppet apply</code> command is
		issued can take several hours to complete.
      </p>
      <small>(<sup>*</sup>We are working on a solution for this.)</small>
      <h3>Updating SUPERSMART</h3>
      <p>
      	The SUPERSMART pipeline is continuously under development. It is therefore wise
      	to periodically check for updates. To do this, boot into the virtual machine
      	(e.g. with <code>vagrant ssh</code>) and issue the following commands:
      </p>
	<pre>
<span class="term">&gt;</span> cd $SUPERSMART_HOME
<span class="term">&gt;</span> git pull</pre>            
      <h3>Advanced configuration</h3>
      <p>
      	The SUPERSMART pipeline is configured using the file 
      	<code>$SUPERSMART_HOME/conf/supersmart.ini</code>. Here are some of the variables
      	in this file that can be configured by users:
      </p>
      <dl>
      	<dt>EXAML_MODEL</dt>
      	<dd>The substitution model used when inferring the backbone phylogeny. Possible 
      	values are GAMMA for a gamma model of rate heterogeneity with 4 discrete rates, 
      	or PSR for a per-site rate category model.</dd>
      	<dt>TREEPL_SMOOTH</dt>
      	<dd>Smoothing factor for the penalized likelihood tree calibration.</dd>
      	<dt>BURNIN</dt>
      	<dd>Fraction of burn-in to omit for the clade inference step.</dd>
      	<dt>MSA_TOOL</dt>
      	<dd>Multiple sequence alignment tool to use. Wrappers are available for a variety
      	of possible aligners (which can be specified here as: mafft, clustalw, kalign, 
      	muscle, probalign, probcons, tcoffee, amap) but the VM comes bundled with mafft
      	and muscle.</dd>
      	<dt>ALN_MERGE</dt>
      	<dd>Specifies the method to use for assigning orthology among aligned clusters.
      	The value 'blast' refers to reciprocal BLAST searches, and is most tested. Other,
      	experimental values are 'inparanoid', where orthology is assigned transitively 
      	from orthology assessments produced by the InParanoid project (for coding regions),
      	or 'feature', which depends on standardized locus names in sequence annotations.      	
      	</dd>
      	<dt>MERGE_OVERLAP</dt>
      	<dd>Specifies the minimum amount of reciprocal overlap among orthologous clusters
      	for them to be merged into a supercluster.</dd>
      	<dt>BACKBONE_INFERENCE_TOOL</dt>
      	<dd>Specifies the tool used for inferring the backbone phylogeny. Possible values
      	are examl or exabayes.</dd>
      	<dt>BACKBONE_MAX_DISTANCE</dt>
      	<dd>Maximum amount of average pairwise sequence distance that the pipeline 
      	accepts for superclusters to be included in the backbone inference step.</dd>
      	<dt>BACKBONE_MIN_COVERAGE</dt>
      	<dd>Minimum number of loci that must be assembled for each taxon in the 
      	supermatrix.</dd>
      	<dt>CLADE_MAX_DISTANCE</dt>
      	<dd>Maximum amount of average pairwise sequence distance that the pipeline 
      	accepts for superclusters to be included in the clade refinement step.</dd>
      	<dt>CLADE_MIN_DENSITY</dt>
      	<dd>Minimum fraction of clade members (taxa) that must be sequenced in an
      	alignment for it to be included in the clade refinement step.</dd>
      	<dt>FOSSIL_BEST_PRACTICE_CUTOFF</dt>
      	<dd>Defines the minimum 'best practice score' for records from the fossil table 
      	to be used as calibration points in the tree calibration.</dd>
      </dl>
      
      <!-- resources section -->
      <h2 id="resources" class="section"><a name="resources"></a>Resources</h2>
		<ul>
			<li><a href="http://github.com/naturalis/supersmart">Source code repository</a></li>
			<li><a href="http://science.naturalis.nl/en/research/bioinformatics/">Naturalis Bioinformatics</a></li>
			<li><a href="http://www.antonelli-lab.net/">Antonelli lab</a></li>
			<li><a href="http://phylota.net">PhyLoTA</a></li>
			<li><a href="http://taxosaurus.org">TaxoSaurus</a></li>
			<li><a href="https://groups.google.com/forum/#!forum/supersmart-dev">SUPERSMART-dev</a></li>
			<li><a href="https://groups.google.com/forum/#!forum/supersmart-users">SUPERSMART-users</a></li>
			<li><a href="https://github.com/naturalis/supersmart/issues">Bug tracker</a></li>
		</ul>      

      <!-- people section -->
      <h2 id="people" class="section"><a name="people"></a>People</h2>
		<ul>
			<li><a href="http://antonelli-lab.net/">Alexandre Antonelli</a></li>
			<li><a href="http://rutgervos.blogspot.com/">Rutger Vos</a></li>
			<li><a href="https://github.com/hettling">Hannes Hettling</a></li>
			<li><a href="http://loco.biosci.arizona.edu/">Mike Sanderson</a></li>
			<li><a href="http://www.bioenv.gu.se/personal/Bengt_Oxelman/">Bengt Oxelman</a></li>
			<li><a href="https://github.com/karinils">Karin Nilsson</a></li>
			<li><a href="http://www.matstopel.se/">Mats Töpel</a></li>
			<li> <a href="http://www.ese.u-psud.fr/article232.html?lang=en">Hervé Sauquet</a></li>
			<li> <a href="http://www.bioenv.gu.se/english/staff/Henrik_Nilsson/">Henrik Nilsson</a></li>
			<li> <a href="http://www2.unil.ch/phylo/member/daniele-silvestro">Daniele Silvestro</a></li>
			<li> <a href="http://scholar.google.com/citations?user=LXR09r0AAAAJ&amp;hl=en">Fabien Condamine</a></li>
			<li><a href="https://github.com/rscharn">Ruud Scharn</a></li>
		</ul> 
	
	  <!-- acknowledgements section -->
      <h2 id="acknowledgements" class="section"><a name="acknowledgements"></a>Contributing organizations</h2>      	  
      <div id="logobox">
      	<a href="http://erc.europa.eu/">
		  <img src="images/erc.png" alt="European Research Council" class="logo" />
		</a>
		<a href="http://www.vr.se/">
		  <img src="images/vetenskapsradet.png" alt="Vetenskapsrådet" class="logo" />
		</a>
		<a href="http://biovel.eu/">
		  <img src="images/biovel.png" alt="BioVeL" class="logo" />
		</a>
		<a href="http://www.gotbot.se/">
		  <img src="images/gothenborg_hortus.png" alt="Göteborgs botaniska trädgård" class="logo" />
		</a>
		<a href="http://www.gu.se/">
		  <img src="images/gothenburg_uni.png" alt="University of Gothenburg" class="logo" />  
		</a>
		<a href="http://www.naturalis.nl/">
		  <img src="images/naturalis.jpg" alt="Naturalis Biodiversity Center" class="logo" /> 
		</a>
		<a href="http://www.uppmax.uu.se/">
		  <img src="images/UPPMAX.png" alt="UPPMAX" class="logo" />                  
		</a>
      </div>

    <!-- FOOTER  -->
    <!--div id="footer_wrap" class="outer">
      <footer class="inner">
        <p class="copyright">SUPERSMART maintained by <a href="https://github.com/naturalis">naturalis</a></p>
        <p>Published with <a href="http://pages.github.com">GitHub Pages</a></p>
      </footer>
    </div-->
    <script type="text/javascript">
      app.initialize();
    </script>
  </body>
</html>
